src/core/jsmemops.h

```diff
@@ -34,7 +34,7 @@
   (BigInt(lower) | (BigInt(upper) << BigInt(32)))
 
 #define IBIGINT_FROM_PAIR(lower, upper)                                        \
-  (BigInt(lower) | (BigInt(upper - 2 * (upper & 0x80000000)) << BigInt(32)))
```

Comment:
Could you add an explanatory comment for this ?

Reply:
Yeah I will add a comment. To quote wikipedia:
https://en.wikipedia.org/wiki/Two%27s_complement#Converting_from_two's_complement_representation

> The value w of an N-bit twos-complement integer ${\displaystyle a_{N-1}a_{N-2}\dots a_{0}} a_{N-1} a_{N-2} \cdots a_0$ is given by the following formula:
>
> $w=-a_{N-1}2^{N-1}+\sum a_{i}2^{i}$

So if we have $v = a_{N-1}2^{N-1}+\sum a_{i}2^{i}$ and we want $w = -a_{N-1}2^{N-1}+\sum a_{i}2^{i}$ we have to subtract $2\cdot a_{N-1}2^{N-1}$. Here $N = 32$. Confusingly, in JavaScript if you take a positive number between $+2^{31}$ and $+2^{32}$ and compute `x & 0x80000000`, the result is $-2^{31}$ so we need to *add* `2*(x & 0x80000000)`.

I don't really understand how bitwise operators work in JavaScript, I always have to use a lot of trial and error. Everything is actually floats so to do bitwise ops it converts from floats to... I'm not sure what size integers and then does the bit ops and then converts everything back to floats. Since there isn't really a leading bit, I don't understand what determines when the result of a bitwise operation is positive or negative. Maybe I should read the ECMAScript spec. The rules for BigInts are equally confusing and also different.

Reply:
Okay further experimentation gives the answer: when you do any bitwise operation, the floats are cast to i32 and then the bitwise operation occurs and then they are cast back to float.  So `(2**32) || 0` is 0, `(2**32 - 1) || 0` is `-1`. JavaScript is weird.

Reply:
yes, it sounds more complicated than I would have expected. 

Should we add some tests for IBIGINT_FROM_PAIR  in JS?